### PR TITLE
Send authenticated remote URL instead of credentials to gitserver when pushing commit

### PR DESCRIPTION
**Pull request by mrnugget**
This is the idea that @keegancsmith had in the feedback on #15554.

Instead of introduction the concept of username/password and token on `gitserver`, we generate an authenticated URL on the client side (the `reconciler` in our case) and send that to `gitserver`.

We construct the URL by using the CloneURL stored on the `repo` and then set the credentials according to which credential we have on hand for which external service.

### CODE REVIEW

File: internal/gitserver/protocol/gitserver.go [hunk 1]
```diff
@@ -197,15 +197,10 @@ type PatchCommitInfo struct {
...
-	//
-	// It's possible to set only the Username field (which is how GitHub
-	// requires its access tokens to be used).
-	//
-	// If both fields are blank, the credentials stored in the remote URL of
-	// the origin in the git repository will be used.
-	Username, Password string
+	// RemoteURL is the git remote URL to which to push the commits.
+	// The URL needs to include HTTP basic auth credentials if no
+	// unauthenticated requests are allowed by the remote host.
```
Comment (eseliger) [inline internal/gitserver/protocol/gitserver.go:202]: Are there any hosts that allow unauthenticated write access?

File: enterprise/internal/campaigns/reconciler.go [hunk 2]
```diff
@@ -170,7 +171,7 @@ func (e *executor) ExecutePlan(ctx context.Context, plan *plan) (err error) {
 	}
 
 	// Set up a source with which we can modify the changeset.
-	e.ccs, err = e.buildChangesetSource(e.repo, extSvc, a)
+	e.ccs, err = e.buildChangesetSource(e.repo, e.extSvc, a)
```
Comment (eseliger) [inline enterprise/internal/campaigns/reconciler.go:174]: Not related to this PR, but I think `buildChangesetSource` should ideally not take any `e.X` arguments, or be a standalone function that doesn't depend on `e` at all.

File: enterprise/internal/campaigns/reconciler_test.go [hunk 3]
```diff
@@ -1464,40 +1467,61 @@ func TestExecutor_UserCredentialsForGitserver(t *testing.T) {
...
+				RemoteURL: "https://fredwoard%20johnssen:my-secret-bbs-token@bitbucket.sourcegraph.com/scm/" + bbsRepo.Name,
+			},
+		},
+		{
+			name: "bitbucketServer site-admin and no credentials",
+			repo: bbsRepo,
+			wantPushConfig: &gitprotocol.PushConfig{
+				RemoteURL: bbsRepoCloneURL,
 			},
 		},
```
Comment (eseliger) [inline enterprise/internal/campaigns/reconciler_test.go:1509]: should we add some `non site-admin` tests, that check for then error being thrown?

**Review by eseliger — APPROVED** [review]
🧹

File: internal/gitserver/protocol/gitserver.go (hunk 1 above)
Comment (mrnugget) [reply internal/gitserver/protocol/gitserver.go:202]: I'm not sure? I just wanted to documented that the URL needs to include that information.

Comment (eseliger) [reply internal/gitserver/protocol/gitserver.go:202]: Yeah, was more a question to myself I guess. No change required :)

File: enterprise/internal/campaigns/reconciler.go (hunk 2 above)
Comment (mrnugget) [reply enterprise/internal/campaigns/reconciler.go:174]: Agree. Can be merged with

https://github.com/sourcegraph/sourcegraph/blob/main/enterprise/internal/campaigns/syncer.go#L548-L565

File: enterprise/internal/campaigns/reconciler_test.go (hunk 3 above)
Comment (mrnugget) [reply enterprise/internal/campaigns/reconciler_test.go:1539]: Added in a6da65b8be

**Review by eseliger — APPROVED** [review]

**Pull request merged**